doc/manual/source/hsms.rst

```diff
@@ -0,0 +1,115 @@
+DNS RRSET data, storing the generated keys in on-disk files.
+
+An Introduction to HSMs
+-----------------------
+
+A hardware security module is typically a tamper proof hardware vault (though
+software variants exist as well) capable of generating and securely storing
+cryptographic keys and performing signing operations using those keys on
+data provided via an interface and returning the signed result via the same
+iinterface.
+
+HSM Interfaces
+--------------
+
+Typically HSMs are interacted with programatically via an interface that
+is compliant with the Oasis PKCS#11 (Public-Key Cryptography Standard)
+specification. Some HSMs also or alternatively support a newer Oasis
+specification called KMIP (Key Management Interoperability Protocol).
+
+KMIP is a data (de)serialization protocol that operates on top of the widely
+used TCP and TLS combination of protocols. As such it requires no additional
+software or special configuration to use and poses no direct security or
+stability threat to the client process.
+
+This is quite different to PKCS#11 which requires the HSM vendor to provide
+a library of code that offers a C language style interface to be used by the
+client at runtime by loading the library (aka module) into its own process
+with no knowledge of or control over what that code is going to do. As
+Cascade is a Rust powered application, crossing the divide between the Rust
+host application and a loaded C library means giving up the stability and
+memory safety guarantees offered by Rust. As such Cascade was designed to
+*NOT* load PKCS#11 modules directly but instead to hand that risk off to a
+helper tool: :program:`kmip2pkcs11`.
+
+Cascade and HSMs
+----------------
+
+Cascade supports both PKCS#11 and KMIP compatible HSMs (initially with KMIP
+support intended primarily for communication with our :program:`kmip2pkcs11` tool,
+though success has also been seen with some other KMIP implementations, e.g.
+Fortanix DSM).
+
+Supported HSMs
+--------------
+
+In principle any HSM supporting PKCS#11 v2.40 or KMIP 1.2 should be supported.
+Note however that for increased security Cascade only supports TLS 1.3
+connections to the KMIP server, even though KMIP 1.2 requires servers to
+offer support for earlier versions of the TLS protocol with known security
+vulnerabilities. For this reason Cascade *CANNOT* be used with PyKMIP as
+PyKMIP only supports older vulnerable TLS ciphers.
+
+Several HSMs have been tested in limited fashion with Cascade. Limited here
+meaning normal usage only, not attempting to deliberately cause problems, and
+not attempting to stress or performance test the interface. The tested HSMs
+are:
+
+- SoftHSM v2.6.1 (PKCS#11, s/w)
+- YubiHSM 2 (PKCS#11, h/w USB key)
+- Thales Cloud HSM (PKCS#11, cloud, v7.3 Cryptovisor 7)
+- Nitrokey NetHSM (PKCS#11, s/w Docker image)
+- Fortanix DSM (KMIP, cloud)
+
+Migration
+---------
+
+How to migrate from one HSM to another can be done in one of three ways:
```

Comment:
Do you want to talk about migrating from on-disk (SoftHSM) to on-disk keys in Cascade? 

A sentence could be added saying something like "Talk to us if you want to migrate from SoftHSM to on-disk keys in Cascade"

(context: https://github.com/NLnetLabs/SoftHSMv2-extractor (not public))